fw/sock_clnt.c

```diff
@@ -428,60 +212,64 @@ tfw_sk_prepare_xmit(struct sock *sk, struct sk_buff *skb, unsigned int mss_now,
 		goto err_purge_tcp_write_queue;
 	}
 
-	*nskbs = UINT_MAX;
+	/*
+	 * This can be possible with very low MTU. Nothing to do, because
+	 * `tcp_write_xmit` break the loop if linit is so small.
+	 */
+	if (unlikely(mss_now < TLS_MAX_OVERHEAD))
+		return 0;
+
 	h2_mode = TFW_CONN_PROTO(conn) == TFW_FSM_H2;
-	if (h2_mode) {
-		r = tfw_h2_sk_prepare_xmit(sk, skb, mss_now, limit, nskbs);
-		if (unlikely(r && r != -ENOMEM))
-			goto err_purge_tcp_write_queue;
-	}
+	if (!h2_mode)
+		return 0;
 
-	return r;
+	h2 = tfw_h2_context(conn);
 
-err_purge_tcp_write_queue:
 	/*
-	 * Leave encrypted segments in the retransmission rb-tree,
-	 * but purge the send queue on unencrypted segments.
+	 * We use two different strategies here:
+	 * - The main strategy is to calculate count of bytes that we can
+	 *   send to the client immediately, according to mss and available
+	 *   TCP CWND. We try to prepare HTTP frame which size is greater
+	 *   or less than mss to prevent HOL problem. The fact that we don't
+	 *   send everything at once allows the higher priority data to be
+	 *   sent before the lower priority data, as we select a new highest
+	 *   priority data to send with each acknowledgement.
+	 * - The second strategy is used when mss is less or equal to frame
+	 *   header size (9 byte, it is possible with very low MTU) or if we
+	 *   finish TCP connection with FIN and should send all available data.
```

Comment:
> if we finish TCP connection with FIN and should send all available data
This approach will be changed in the future, please update comment when the changes are made.